packages/kbn-handlebars/index.ts

```diff
@@ -623,70 +646,79 @@ class ElasticHandlebarsVisitor extends Handlebars.Visitor {
+      params: this.resolveNodes(node.params),
+      options: this.setupParams(node, helperName),
     };
   }
 
-  private setupDecoratorOptions(
-    decorator: hbs.AST.Decorator | hbs.AST.DecoratorBlock
-  ): Handlebars.HelperOptions {
+  private setupDecoratorOptions(decorator: hbs.AST.Decorator | hbs.AST.DecoratorBlock) {
     // TypeScript: The types indicate that `decorator.path` technically can be an `hbs.AST.Literal`. However, the upstream codebase always treats it as an `hbs.AST.PathExpression`, so we do too.
     const name = (decorator.path as hbs.AST.PathExpression).original;
-    const options = this.setupParams(decorator as hbs.AST.DecoratorBlock, name);
+    const options = toDecoratorOptions(this.setupParams(decorator, name));
 
     if (decorator.params.length > 0) {
       if (!this.processedRootDecorators) {
         // When processing the root decorators, temporarily remove the root context so it's not accessible to the decorator
         const context = this.scopes.shift();
-        // @ts-expect-error: Property 'args' does not exist on type 'HelperOptions'. The 'args' property is expected in decorators
         options.args = this.resolveNodes(decorator.params);
         this.scopes.unshift(context);
       } else {
-        // @ts-expect-error: Property 'args' does not exist on type 'HelperOptions'. The 'args' property is expected in decorators
         options.args = this.resolveNodes(decorator.params);
       }
     }
 
-    // @ts-expect-error: Property 'lookupProperty' does not exist on type 'HelperOptions'
-    delete options.lookupProperty; // There's really no tests/documentation on this, but to match the upstream codebase we'll remove `lookupProperty` from the decorator context
-
     return options;
   }
 
-  private setupParams(
-    node: ProcessableNodeWithPathParts,
-    helperName: string
-  ): Handlebars.HelperOptions {
-    const options: Handlebars.HelperOptions = {
-      // @ts-expect-error: Name should be on there, but the offical types doesn't know this
-      name: helperName,
+  private setupParams(node: ProcessableBlockStatementNode, name: string): Handlebars.HelperOptions;
+  private setupParams(node: ProcessableStatementNode, name: string): NonBlockHelperOptions;
+  private setupParams(node: ProcessableNode, name: string): AmbiguousHelperOptions;
+  private setupParams(node: ProcessableNode, name: string): AmbiguousHelperOptions {
+    // TODO: Is this the way to type `options` based on `node`?
+    type OptionsReturnType = typeof node extends hbs.AST.BlockStatement
+      ? Handlebars.HelperOptions
+      : NonBlockHelperOptions;
+    // TODO: I have to use `Partial` because `lookupProperty` (which is a required property of `OptionsReturnType`) isn't defined when the `options` object is first constructed.
+    // But it's added via the `Object.assign` at the end. But is this way to do make TypeScript happy in this scenario?
```

Comment:
`this.defaultHelperOptions` does not contain a required `lookupProperty` so I'm not sure if it's guaranteed to be added there

Reply:
Yeah it's a bit hard to reason about. It's added programatically at runtime in the constructor of the `ElasticHandlebarsVisitor` class. So it's always there.

Writing this comment actually gave me an idea that can make it a lot better/easier to read. I'll push an update...